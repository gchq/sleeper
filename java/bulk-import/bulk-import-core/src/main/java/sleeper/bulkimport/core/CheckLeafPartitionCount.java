/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.bulkimport.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.bulkimport.core.job.BulkImportJob;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.StateStore;

import static sleeper.core.properties.table.TableProperty.BULK_IMPORT_MIN_LEAF_PARTITION_COUNT;

public class CheckLeafPartitionCount {
    private static final Logger LOGGER = LoggerFactory.getLogger(CheckLeafPartitionCount.class);

    private CheckLeafPartitionCount() {
    }

    public static boolean hasMinimumPartitions(
            TableProperties tableProperties, StateStore stateStore, BulkImportJob job) {
        int leafPartitionCount = stateStore.getLeafPartitions().size();
        int minPartitionCount = tableProperties.getInt(BULK_IMPORT_MIN_LEAF_PARTITION_COUNT);
        if (leafPartitionCount < minPartitionCount) {
            LOGGER.info("Minimum partition count was {}, but found {} leaf partitions. Skipping job {}",
                    minPartitionCount, leafPartitionCount, job.getId());
            return false;
        } else {
            LOGGER.info("Minimum partition count has been met. Running job {}", job.getId());
            return true;
        }
    }
}
