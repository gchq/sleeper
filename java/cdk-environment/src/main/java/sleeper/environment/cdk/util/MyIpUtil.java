/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.environment.cdk.util;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class MyIpUtil {

    private MyIpUtil() {
        // Prevent instantiation
    }

    public static String findMyIp() {
        try {
            URL url = new URL("https://checkip.amazonaws.com/");
            try (BufferedReader br = new BufferedReader(
                    new InputStreamReader(url.openStream(), StandardCharsets.UTF_8))) {
                return br.readLine();
            }
        } catch (IOException e) {
            throw new RuntimeException("Could not find my IP", e);
        }
    }
}
