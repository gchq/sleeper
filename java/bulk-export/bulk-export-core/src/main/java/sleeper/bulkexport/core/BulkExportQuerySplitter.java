/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.bulkexport.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.bulkexport.core.model.BulkExportLeafPartitionQuery;
import sleeper.bulkexport.core.model.BulkExportQuery;
import sleeper.core.partition.Partition;
import sleeper.core.partition.PartitionTree;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.StateStore;
import sleeper.core.statestore.StateStoreException;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static sleeper.core.properties.table.TableProperty.QUERY_PROCESSOR_CACHE_TIMEOUT;

/**
 * Splits up an export query into leaf partition export queries.
 */
public class BulkExportQuerySplitter {
    private static final Logger LOGGER = LoggerFactory.getLogger(BulkExportQuerySplitter.class);

    private final StateStore stateStore;
    private final TableProperties tableProperties;
    private final Supplier<String> idSupplier;
    private final Supplier<Instant> timeSupplier;
    private List<Partition> leafPartitions;
    private PartitionTree partitionTree;
    private Map<String, List<String>> partitionToFiles;
    private Instant nextInitialiseTime;

    public BulkExportQuerySplitter(TableProperties tableProperties, StateStore stateStore) {
        this(stateStore, tableProperties, () -> UUID.randomUUID().toString(), Instant::now);
    }

    public BulkExportQuerySplitter(StateStore stateStore, TableProperties tableProperties, Supplier<String> idSupplier, Supplier<Instant> timeSupplier) {
        this.stateStore = stateStore;
        this.tableProperties = tableProperties;
        this.idSupplier = idSupplier;
        this.timeSupplier = timeSupplier;
        init(timeSupplier.get());
    }

    /**
     * Initialises the partitions and the mapping from partitions to active files if
     * needed.
     *
     * @throws StateStoreException if the state store can't be accessed
     */
    public void initIfNeeded() throws StateStoreException {
        Instant now = timeSupplier.get();
        if (nextInitialiseTime.isAfter(now)) {
            LOGGER.debug("Not refreshing state for table {}", tableProperties.getStatus());
            return;
        }
        init(now);
    }

    private void init(Instant now) throws StateStoreException {
        List<Partition> partitions = stateStore.getAllPartitions();
        Map<String, List<String>> partitionToFileMapping = stateStore.getPartitionToReferencedFilesMap();
        init(partitions, partitionToFileMapping, now);
    }

    private void init(List<Partition> partitions, Map<String, List<String>> partitionToFileMapping, Instant now) {
        leafPartitions = partitions.stream()
                .filter(Partition::isLeafPartition)
                .collect(Collectors.toList());
        partitionTree = new PartitionTree(partitions);
        partitionToFiles = partitionToFileMapping;
        nextInitialiseTime = now.plus(tableProperties.getInt(QUERY_PROCESSOR_CACHE_TIMEOUT), ChronoUnit.SECONDS);
        LOGGER.info("Loaded state for table {}. Found {} partitions. Next initialise time: {}",
                tableProperties.getStatus(), partitions.size(), nextInitialiseTime);
    }

    /**
     * Splits up a query into a sub-query per relevant leaf partition. For each leaf partition, it
     * finds the parent partitions in the tree and adds any files still belonging
     * to the parent to the sub query.
     *
     * @param  bulkExportQuery the query to be split up
     * @return                 a list of {@link LeafPartitionQuery}s
     */
    public List<BulkExportLeafPartitionQuery> splitIntoLeafPartitionQueries(BulkExportQuery bulkExportQuery) {
        LOGGER.debug("There are {} relevant leaf partitions", leafPartitions.size());

        List<BulkExportLeafPartitionQuery> leafPartitionQueriesList = new ArrayList<>();
        for (Partition partition : leafPartitions) {
            List<String> files = getFiles(partition);

            if (files.isEmpty()) {
                LOGGER.info("No files for partition {}", partition.getId());
                continue;
            }

            // For each leaf partition, create query with pre-populated list of files that
            // need to be read.
            BulkExportLeafPartitionQuery bulkExportLeafPartitionQuery = BulkExportLeafPartitionQuery
                    .forPartition(bulkExportQuery, tableProperties, partition)
                    .subExportId(idSupplier.get())
                    .files(files)
                    .build();
            LOGGER.debug("Created {}", bulkExportLeafPartitionQuery);
            leafPartitionQueriesList.add(bulkExportLeafPartitionQuery);
        }

        return leafPartitionQueriesList;
    }

    /**
     * Gets a list of file paths for a partition.
     *
     * @param  partition which partition to get the files from
     * @return           a list of files
     */
    protected List<String> getFiles(Partition partition) {
        // Get all partitions up to the root of the tree
        List<String> relevantPartitions = new ArrayList<>();
        relevantPartitions.add(partition.getId());
        relevantPartitions.addAll(partitionTree.getAllAncestorIds(partition.getId()));

        // Get relevant files
        List<String> files = new ArrayList<>();
        for (String partitionId : relevantPartitions) {
            List<String> filesForPartition = partitionToFiles.get(partitionId);
            if (null != filesForPartition) {
                files.addAll(partitionToFiles.get(partitionId));
            }
        }
        return files;
    }
}
