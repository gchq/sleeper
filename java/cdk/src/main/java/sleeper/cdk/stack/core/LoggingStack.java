/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.cdk.stack.core;

import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.services.logs.ILogGroup;
import software.amazon.awscdk.services.logs.LogGroup;
import software.constructs.Construct;

import sleeper.cdk.util.Utils;
import sleeper.core.properties.instance.InstanceProperties;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static sleeper.core.properties.instance.CommonProperty.LOG_RETENTION_IN_DAYS;

public class LoggingStack extends NestedStack {

    private final Map<LogGroupRef, ILogGroup> logGroupByName = new HashMap<>();
    private final InstanceProperties instanceProperties;

    public LoggingStack(Construct scope, String id, InstanceProperties instanceProperties) {
        super(scope, id);
        this.instanceProperties = instanceProperties;

        for (LogGroupRef ref : LogGroupRef.values()) {
            createLogGroup(ref);
        }
    }

    public ILogGroup getLogGroup(LogGroupRef ref) {
        return Objects.requireNonNull(logGroupByName.get(ref), "No log group found: " + ref);
    }

    private void createLogGroup(LogGroupRef ref) {
        logGroupByName.put(ref, LogGroup.Builder.create(this, ref.shortName)
                .logGroupName(ref.prefix + String.join("-", "sleeper", Utils.cleanInstanceId(instanceProperties), ref.shortName))
                .retention(Utils.getRetentionDays(instanceProperties.getInt(LOG_RETENTION_IN_DAYS)))
                .build());
    }

    public enum LogGroupRef {

        // Accessed directly by getter on this class
        VPC_CHECK("vpc-check"),
        VPC_CHECK_PROVIDER("vpc-check-provider"),
        CONFIG_AUTODELETE("config-autodelete"),
        CONFIG_AUTODELETE_PROVIDER("config-autodelete-provider"),
        TABLE_DATA_AUTODELETE("table-data-autodelete"),
        TABLE_DATA_AUTODELETE_PROVIDER("table-data-autodelete-provider"),
        STATESTORE_COMMITTER("statestore-committer"),

        // Accessed via CoreStacks getters
        PROPERTIES_WRITER("properties-writer"),
        PROPERTIES_WRITER_PROVIDER("properties-writer-provider"),
        STATE_SNAPSHOT_CREATION_TRIGGER("state-snapshot-creation-trigger"),
        STATE_SNAPSHOT_CREATION("state-snapshot-creation"),
        STATE_SNAPSHOT_DELETION_TRIGGER("state-snapshot-deletion-trigger"),
        STATE_SNAPSHOT_DELETION("state-snapshot-deletion"),
        STATE_TRANSACTION_DELETION_TRIGGER("state-transaction-deletion-trigger"),
        STATE_TRANSACTION_DELETION("state-transaction-deletion"),
        METRICS_TRIGGER("metrics-trigger"),
        METRICS_PUBLISHER("metrics-publisher"),
        BULK_IMPORT_EMR_SERVERLESS_START("bulk-import-EMRServerless-start"),
        BULK_IMPORT_EMR_NON_PERSISTENT_START("bulk-import-NonPersistentEMR-start"),
        BULK_IMPORT_EMR_PERSISTENT_START("bulk-import-PersistentEMR-start"),
        BULK_IMPORT_EKS_STARTER("bulk-import-eks-starter"),
        BULK_IMPORT_EKS("bulk-import-eks"),
        BULK_IMPORT_EKS_STATE_MACHINE("EksBulkImportStateMachine", "/aws/vendedlogs/states/"),
        BULK_IMPORT_AUTODELETE("bulk-import-autodelete"),
        BULK_IMPORT_AUTODELETE_PROVIDER("bulk-import-autodelete-provider"),
        INGEST_TASKS("IngestTasks"),
        INGEST_CREATE_TASKS("ingest-create-tasks"),
        INGEST_BATCHER_SUBMIT_FILES("ingest-batcher-submit-files"),
        INGEST_BATCHER_CREATE_JOBS("ingest-batcher-create-jobs"),
        PARTITION_SPLITTING_TRIGGER("partition-splitting-trigger"),
        PARTITION_SPLITTING_FIND_TO_SPLIT("partition-splitting-find-to-split"),
        PARTITION_SPLITTING_HANDLER("partition-splitting-handler"),
        COMPACTION_TASKS_FARGATE("FargateCompactionTasks"),
        COMPACTION_TASKS_EC2("EC2CompactionTasks"),
        COMPACTION_JOB_CREATION_TRIGGER("compaction-job-creation-trigger"),
        COMPACTION_JOB_CREATION_HANDLER("compaction-job-creation-handler"),
        COMPACTION_JOB_DISPATCHER("compaction-job-dispatcher"),
        COMPACTION_TASKS_CREATOR("compaction-tasks-creator"),
        COMPACTION_CUSTOM_TERMINATION("compaction-custom-termination"),
        GARBAGE_COLLECTOR_TRIGGER("garbage-collector-trigger"),
        GARBAGE_COLLECTOR("garbage-collector"),
        QUERY_EXECUTOR("query-executor"),
        QUERY_LEAF_PARTITION("query-leaf-partition"),
        QUERY_WEBSOCKET_HANDLER("query-websocket-handler"),
        QUERY_RESULTS_AUTODELETE("query-results-autodelete"),
        QUERY_RESULTS_AUTODELETE_PROVIDER("query-results-autodelete-provider"),
        QUERY_KEEP_WARM("query-keep-warm"),
        SIMPLE_ATHENA_HANDLER("Simple-athena-handler"),
        ITERATOR_APPLYING_ATHENA_HANDLER("IteratorApplying-athena-handler"),
        SPILL_BUCKET_AUTODELETE("spill-bucket-autodelete"),
        SPILL_BUCKET_AUTODELETE_PROVIDER("spill-bucket-autodelete-provider");

        private final String shortName;
        private final String prefix;

        LogGroupRef(String shortName) {
            this(shortName, "");
        }

        LogGroupRef(String shortName, String prefix) {
            this.shortName = shortName;
            this.prefix = prefix;
        }
    }
}
