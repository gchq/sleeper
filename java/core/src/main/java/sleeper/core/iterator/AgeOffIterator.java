/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.iterator;

import sleeper.core.iterator.closeable.CloseableIterator;
import sleeper.core.iterator.closeable.FilteringIterator;
import sleeper.core.row.Row;
import sleeper.core.schema.Schema;

import java.util.List;

/**
 * Filters out rows older than a specified age. If the specified timestamp field is more than a certain length of
 * time ago then the row is removed. This is an example implementation of {@link ConfigStringIterator}.
 */
public class AgeOffIterator implements ConfigStringIterator {
    private AgeOffFilter filter;

    public AgeOffIterator() {
    }

    public AgeOffIterator(AgeOffFilter filter) {
        this.filter = filter;
    }

    @Override
    public void init(String configString, Schema schema) {
        String[] fields = configString.split(",");
        if (2 != fields.length) {
            throw new IllegalArgumentException("Configuration string should have 2 fields: field name and age off time");
        }
        filter = new AgeOffFilter(fields[0], Long.parseLong(fields[1]));
    }

    @Override
    public List<String> getRequiredValueFields() {
        return List.of(filter.getFieldName());
    }

    @Override
    public CloseableIterator<Row> apply(CloseableIterator<Row> input) {
        return new FilteringIterator<>(input, filter);
    }
}
