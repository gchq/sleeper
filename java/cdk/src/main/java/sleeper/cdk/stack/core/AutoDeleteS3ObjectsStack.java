/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.cdk.stack.core;

import software.amazon.awscdk.CustomResource;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.customresources.Provider;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.logs.ILogGroup;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.IBucket;
import software.constructs.Construct;

import sleeper.cdk.jars.BuiltJars;
import sleeper.cdk.jars.LambdaCode;
import sleeper.cdk.stack.core.LoggingStack.LogGroupRef;
import sleeper.cdk.util.Utils;
import sleeper.core.deploy.LambdaHandler;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.util.EnvironmentUtils;

import java.util.Map;

public class AutoDeleteS3ObjectsStack extends NestedStack {

    private IFunction lambda;
    private Provider provider;

    public AutoDeleteS3ObjectsStack(Construct scope, String id, InstanceProperties instanceProperties, BuiltJars jars, LoggingStack loggingStack) {
        super(scope, id);
        createLambda(id, instanceProperties, jars, loggingStack.getLogGroup(LogGroupRef.AUTO_DELETE_S3_OBJECTS),
                loggingStack.getLogGroup(LogGroupRef.AUTO_DELETE_S3_OBJECTS_PROVIDER));
    }

    public AutoDeleteS3ObjectsStack(Construct scope, String id, InstanceProperties instanceProperties, BuiltJars jars, Integer retentionDays) {
        super(scope, id);
        ILogGroup logGroup = LogGroup.Builder.create(this, id + "-AutoDeleteLambdaLogGroup")
                .logGroupName("s3-bucket-autodelete")
                .retention(Utils.getRetentionDays(retentionDays))
                .build();
        ILogGroup providerLogGroup = LogGroup.Builder.create(this, id + "-AutoDeleteProviderLogGroup")
                .logGroupName("s3-bucket-autodelete-provider")
                .retention(Utils.getRetentionDays(retentionDays))
                .build();
        createLambda(id, instanceProperties, jars, logGroup, providerLogGroup);
    }

    private void createLambda(String id, InstanceProperties instanceProperties, BuiltJars jars, ILogGroup logGroup, ILogGroup providerLogGroup) {

        // Jars bucket
        IBucket jarsBucket = Bucket.fromBucketName(this, "JarsBucket", jars.bucketName());
        LambdaCode lambdaCode = jars.lambdaCode(jarsBucket);

        String functionName = String.join("-", "sleeper",
                Utils.cleanInstanceId(instanceProperties), "auto-delete-s3-objects");

        lambda = lambdaCode.buildFunction(this, LambdaHandler.AUTO_DELETE_S3_OBJECTS, id + "Lambda", builder -> builder
                .functionName(functionName)
                .memorySize(2048)
                .environment(EnvironmentUtils.createDefaultEnvironmentNoConfigBucket(instanceProperties))
                .description("Lambda for auto-deleting S3 objects")
                .logGroup(logGroup)
                .timeout(Duration.minutes(10)));

        provider = Provider.Builder.create(this, id + "Provider")
                .onEventHandler(lambda)
                .logGroup(providerLogGroup)
                .build();
    }

    public void grantAccessToCustomResource(String id, InstanceProperties instanceProperties,
            IBucket bucket, String bucketName) {

        bucket.grantRead(lambda);
        bucket.grantDelete(lambda);

        CustomResource.Builder.create(this, id + "-AutoDelete")
                .resourceType("Custom::AutoDeleteS3Objects")
                .properties(Map.of("bucket", bucketName))
                .serviceToken(provider.getServiceToken())
                .build();
    }

}
