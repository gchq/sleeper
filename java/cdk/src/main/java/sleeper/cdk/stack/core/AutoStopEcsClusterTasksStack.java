/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.cdk.stack.core;

import software.amazon.awscdk.CustomResource;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.customresources.Provider;
import software.amazon.awscdk.services.iam.IRole;
import software.amazon.awscdk.services.iam.ManagedPolicy;
import software.amazon.awscdk.services.iam.PolicyStatement;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.IBucket;
import software.constructs.Construct;

import sleeper.cdk.jars.BuiltJars;
import sleeper.cdk.jars.LambdaCode;
import sleeper.cdk.stack.core.LoggingStack.LogGroupRef;
import sleeper.cdk.util.Utils;
import sleeper.core.deploy.LambdaHandler;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.util.EnvironmentUtils;

import java.util.List;
import java.util.Map;
import java.util.Objects;

public class AutoStopEcsClusterTasksStack extends NestedStack {

    public AutoStopEcsClusterTasksStack(Construct scope, String id, InstanceProperties instanceProperties, BuiltJars jars,
            NestedStack clusterStack, LoggingStack logging) {
        super(scope, id);

        // Jars bucket
        IBucket jarsBucket = Bucket.fromBucketName(this, "JarsBucket", jars.bucketName());
        LambdaCode lambdaCode = jars.lambdaCode(jarsBucket);

        String functionName = String.join("-", "sleeper",
                Utils.cleanInstanceId(instanceProperties), "auto-stop-ecs-cluster-tasks");

        IFunction lambda = lambdaCode.buildFunction(this, LambdaHandler.AUTO_STOP_ECS_CLUSTER_TASKS, "AutoStopEcsClusterTasksLambda", builder -> builder
                .functionName(functionName)
                .memorySize(2048)
                .environment(EnvironmentUtils.createDefaultEnvironmentNoConfigBucket(instanceProperties))
                .description("Lambda for auto-stopping ECS tasks")
                .logGroup(logging.getLogGroup(LogGroupRef.ECS_CLUSTER_TASKS_AUTOSTOP))
                .timeout(Duration.minutes(10)));

        // Grant this function permission to list tasks and stop tasks
        PolicyStatement policyStatement = PolicyStatement.Builder
                .create()
                .resources(List.of("*"))
                .actions(List.of("ecs:ListTasks", "ecs:StopTask", "iam:PassRole"))
                .build();
        IRole role = Objects.requireNonNull(lambda.getRole());
        role.addToPrincipalPolicy(policyStatement);
        role.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName("service-role/AmazonECSTaskExecutionRolePolicy"));

        Provider propertiesWriterProvider = Provider.Builder.create(scope, id + "Provider")
                .onEventHandler(lambda)
                .logGroup(logging.getLogGroup(LogGroupRef.ECS_CLUSTER_TASKS_AUTOSTOP_PROVIDER))
                .build();

        CustomResource customResource = CustomResource.Builder.create(scope, id)
                .resourceType("Custom::AutoStopEcsClusterTasks")
                .properties(Map.of("cluster", clusterStack.getNode().getId()))
                .serviceToken(propertiesWriterProvider.getServiceToken())
                .build();

        customResource.getNode().addDependency(clusterStack);

        Utils.addStackTagIfSet(this, instanceProperties);

    }

}
