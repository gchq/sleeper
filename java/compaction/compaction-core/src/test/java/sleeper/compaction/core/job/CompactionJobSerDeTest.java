/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;

import sleeper.core.iterator.AgeOffIterator;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class CompactionJobSerDeTest {

    private final CompactionJobSerDe serDe = new CompactionJobSerDe();

    @Test
    void shouldConvertCompactionJobToFromJsonWithNoIterator() {
        // Given
        CompactionJob job = CompactionJob.builder()
                .tableId("test-table")
                .jobId("test-job")
                .inputFiles(Arrays.asList("file1", "file2"))
                .outputFile("outputfile")
                .partitionId("test-partition")
                .build();

        // When
        String json = serDe.toJsonPrettyPrint(job);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(job);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldConvertCompactionJobToFromJsonWithIterator() {
        // Given
        CompactionJob job = CompactionJob.builder()
                .tableId("test-table")
                .jobId("test-job")
                .inputFiles(Arrays.asList("file1", "file2"))
                .outputFile("outputfile")
                .partitionId("test-partition")
                .iteratorClassName(AgeOffIterator.class.getName())
                .iteratorConfig("key,600000")
                .build();

        // When
        String json = serDe.toJsonPrettyPrint(job);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(job);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldConvertCompactionJobBatchToFromJson() {
        // Given
        CompactionJob job1 = CompactionJob.builder()
                .tableId("some-table")
                .jobId("some-job")
                .inputFiles(Arrays.asList("file1", "file2"))
                .outputFile("outputfile1")
                .partitionId("some-partition")
                .build();
        CompactionJob job2 = CompactionJob.builder()
                .tableId("other-table")
                .jobId("other-job")
                .inputFiles(Arrays.asList("file3", "file4"))
                .outputFile("outputfile2")
                .partitionId("other-partition")
                .iteratorClassName(AgeOffIterator.class.getName())
                .iteratorConfig("key,600000")
                .build();

        // When
        String json = serDe.toJsonPrettyPrint(List.of(job1, job2));

        // Then
        assertThat(serDe.batchFromJson(json)).containsExactly(job1, job2);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }
}
