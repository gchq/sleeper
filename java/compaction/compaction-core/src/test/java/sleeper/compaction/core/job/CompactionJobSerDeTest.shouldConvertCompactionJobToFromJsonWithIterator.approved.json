{
  "tableId": "test-table",
  "jobId": "test-job",
  "inputFiles": [
    "file1",
    "file2"
  ],
  "outputFile": "outputfile",
  "partitionId": "test-partition",
  "iteratorClassName": "sleeper.example.iterator.AgeOffIterator",
  "iteratorConfig": "key,600000"
}